### PR TITLE
Feature/adduser-group

### PR DESCRIPTION
**Pull request by mahpatil**
This pull request adds support for creating users, groups and user group mappings.
In addition, Makefile has been included to build and test locally

### CODE REVIEW

**Comment by tazjin** [comment]
Hey,

thanks for submitting this - as you mentioned in #23, I don't actively maintain this project. We've switched from Keycloak to a more straightforward, custom implementation at work.

If you're interested, you could take over ownership of the project and I'd point people at your fork from my `README`. There are a couple of other users (cc: @ggulati2, @evilezh, @phillipj) who may also be interested in the future of the project - but I'd prefer to stay out of it.

Cheers!

**Comment by phillipj** [comment]
Just to chime in, we're definitively still using this and interested in contributing more going forward.

**Comment by tazjin** [comment]
@phillipj Same comment regarding ownership for you, also - if you (or Tomra) would like to be the maintainer!

**Comment by mahpatil** [comment]
@tazjin sure. if you want to transfer the ownerhip, would like to include the features to the community

**Comment by mahpatil** [comment]
@tazjin, @phillipj can you let me know? If that isnt going to happen, I can create a separate repo and merge my own PR

**Comment by phillipj** [comment]
Sorry this slipped through the cracks and got lost in my never ending backlog..

Great you got to merge this, managing users is really valuable!

File: src/keycloak/client.go [hunk 1]
```diff
@@ -91,6 +91,7 @@ func (c *KeycloakClient) DeleteClient(id string, realm string) error {
 
 func (c *KeycloakClient) GetClientServiceAccountUser(id string, realm string) (*User, error) {
 	url := fmt.Sprintf(clientUserUri, c.url, realm, id)
+	fmt.Println("Hi")
```
Comment (phillipj) [inline src/keycloak/client.go:94]: This should probably not be permanent?

File: src/keycloak/group_mappings.go [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+	if err != nil {
+		return nil, err
+	}
+
+	ug := UserGroupMap{}
+	ug.GroupId = groupId
+	ug.Realm = realm
+	ug.UserIds = []string{}
+	log.Printf(string(len(users)))
```
Comment (phillipj) [inline src/keycloak/group_mappings.go:32]: Do we need to log the length of users here?

File: test/main.tf [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+  realm     = "Jenkins"
+  client_id = "test"
+  redirect_uris = ["http://127.0.0.1:8080/"]
+  service_accounts_enabled = true
+}
+*/
+
+resource "keycloak_group" "group1" {
+  name       = "group1"
+  realm      = "Jenkins"
```
Comment (phillipj) [inline test/main.tf:12]: Did you consider referencing terraform resource ids or similar here, rather than static realm names?

**Comment by tazjin** [comment]
@phillipj @wonoments I gave both of you commit access to the repository, it's kind of up to you how you want to deal with it. If you decide to move the main repository somewhere else, feel free to add a note to the README.

**Comment by phillipj** [comment]
Thanks @tazjin 👍

File: src/keycloak/client.go (hunk 1 above)
Comment (mahpatil) [reply src/keycloak/client.go:94]: Thanks @phillipj will resolve in the latest PR

File: src/keycloak/group_mappings.go (hunk 2 above)
Comment (mahpatil) [reply src/keycloak/group_mappings.go:32]: Thanks @phillipj will resolve in the latest PR

**Pull request merged**